### PR TITLE
Change links to buttons, expose `aria-pressed` for toggles, add `aria-expanded` to send view's "Options"

### PR DESCRIPTION
**Pull request by patrickhlauke**
- also make existing `<a routerLink...>` type controls keyboard focusable with the addition of `tabindex="0"`

## Type of change

- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

- Make sure that controls that act as buttons are exposed correctly to assistive technologies as buttons (either by changing them from links to `<button>` elements, or making sure they have an appropriate `role="button"`.
- For links with `role="button"`, also make sure that they're focusable - if they don't have an `href="#"`, add a `tabindex="0"` - only left `routerLink` links that way, as I'm not familiar enough with angular to know if this would otherwise break functionality if changed to a proper `<button>`. Ideally though, if they act as buttons, there's no reason not to also make them `<button>` elements, they just need some amended styles.
- All link that have `<a href="#" role="button" ...>` have been changed to actual `<button ...>` elements - both because that's cleaner code, and because traditionally controls announced as "button" by assistive technologies can then be activated both using `Enter` or `Space`, but in this case the `<a href="#" role="button" ...>` controls only currently react to `Enter`. Now that they're `<button ...>` elements, they can be consistently activated (with exception of the `routerLink` ones mentioned above).
- For controls in the left-hand column, add `aria-pressed` and set it to the appropriate value (`true`/`false`) to programmatically convey when a category is actually pressed/selected/active.
- Add `aria-expanded` to a stray expand/collapse control (in the Send view) that lacked it.
- Fix "Options" button in send view (lacked accessible name because "Options" text was outside of the control), add `aria-expanded`
- Minor style tweaks/fixes - in particular, fix off-center outline on last button in any action-button group

Closes https://github.com/bitwarden/desktop/issues/1435
Closes https://github.com/bitwarden/desktop/issues/1439

See also https://github.com/bitwarden/browser/pull/1988

## Testing requirements

- test using a screen reader (e.g. NVDA on Windows)
- although I think I've gone through most screens/dialogs to ensure that changes from `<a ...>` to `<button ...>` remain visually the same, there are some dialogs that I don't have access to (as I only have a free account, so don't see things like two-factor authentication specific things, or ability to define collections/nested folders it seems?). these will need to be very carefully compared to the current stable version to make sure they look the same after the CSS and code changes here.

## Before you submit

- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by patrickhlauke** [comment]
Posting this as a draft request for now, as I'm unable to get the build environment set up on my machine. admittedly, I'm fairly low skill when it comes to build envs, and angular, so likely this will require further tweaks/corrections. Hoping that somebody from the team (@addisonbeck @djsmith85 ?) can point me in the right direction / help me shepherd this through to a proper merge-ready state

**Comment by Hinton** [comment]
Hi @patrickhlauke,

Mind expanding on what build issues you are encountering? 

Building on windows is a bit tricky since you need VS 2019 or 2022 with C++ build tools installed + 2015 build tools, and the Windows 10 SDK 17134, which you can download from https://developer.microsoft.com/en-us/windows/downloads/sdk-archive/.

**Comment by patrickhlauke** [comment]
@hinton ah, i think i worked out what i was doing wrong. i've got vscode set up to use the windows linux subsystem install of ubuntu as my terminal i regularly use for build stuff. i had installed the microsoft things (I think I got them all right), but ubuntu was obviously not seeing them. i'll have a try at installing everything on the windows side and using a windows command line or powershell to build

**Comment by patrickhlauke** [comment]
@Hinton just to check, for the VS 2022 installer, are these the extra packages? (that installer is extremely confusing...)

installer with listed packages: windows universal c runtime, c++ build tools core features, windows universal crt sdk, mscv v140 vs 2015 c++ build tools

**Comment by Hinton** [comment]
Ah yes, you will need to have node setup directly on windows. And run `npm run electron` on directly on the windows side as well.

I'm a bit unsure what the minimum available components would be, I have `Desktop development with C++` installed as well.
You will also need the "Windows 10 SDK, version 1803 (10.0.17134.12)" from the archive link above, since the newer SDKs are not compatible with one of the dependencies.

**Comment by patrickhlauke** [comment]
thanks @Hinton ... i now have a working build environment - my minimal VS setup above, plus the SDK, node, and oddly enough python, did the job in the end, and then running it from powershell. i'm still sorting out a few of the styling issues that arise from the change from `<a>` to `<button>`, but should have this PR complete and ready for review soon.

**Comment by Hinton** [comment]
Python ... Yea, Much of the node land uses `node-gyp` which is written in python.

**Comment by patrickhlauke** [comment]
There are two visual aspects that I have not been able to address in this PR.

In the notification toasts, when I test locally, the "X" times character for the close control looks embossed for some reason. I was not able to determine what would cause this, as I don't *think* any of the CSS changes I've made would affect this.

toast notification, with the embossed-looking 'x' close control

Secondly, in the "Export vault" dialog, the layout is messed up. However, it also currently seems messed up the same way if I just run the current `master` from this repo, so seems to be a problem in the current master branch as well, not related to this PR.


export vault dialog, showing the messed up layout

**Comment by Hinton** [comment]
Thanks for the headsup of the export component, I have a pending fix in jslib https://github.com/bitwarden/jslib/pull/747 which will also need fixes in the clients.

**Comment by patrickhlauke** [comment]
Need to solve the recent merge conflict here too. I'm on the case.

**Review by djsmith85 — CHANGES_REQUESTED** [review]
Excellent work @patrickhlauke as always!!  Thank you for being patient with the review.

Just some minor things and if you are up for it 2 additional changes

1. Convert the send items in the send list also into buttons
2. With the latest release, we included a user-name generator which is a link instead of button.

Feel free to let me know if this is out of scope and I'll take care of it.

**Comment by patrickhlauke** [comment]
thanks for reviewing @djsmith85 ... i'll have a look and tackle your comments in the coming days :+1:

**Comment by patrickhlauke** [comment]
@djsmith85 on your first point

> 1 Convert the send items in the send list also into buttons

Do you mean this (which I already did in this PR) https://github.com/bitwarden/desktop/pull/1437/files#diff-1f721a7ba41f87261672da35a0d193a9079c3894eecfc8646d0cbaf18083b543 ?

**Comment by djsmith85** [comment]
> @djsmith85 on your first point
> 
> > 1 Convert the send items in the send list also into buttons
> 
> Do you mean this (which I already did in this PR) https://github.com/bitwarden/desktop/pull/1437/files#diff-1f721a7ba41f87261672da35a0d193a9079c3894eecfc8646d0cbaf18083b543 ?

Almost, you did the filters on the left hand side. I meant the actual list of [sends](https://github.com/bitwarden/desktop/blob/b46720644839b246a5c6587f17e3df0f49f54b51/src/app/send/send.component.html#L39)

**Comment by patrickhlauke** [comment]
@djsmith85 aha, thanks. goes to show i never used sends before :)

**Comment by patrickhlauke** [comment]
@djsmith85 on the second point

> 2 With the latest release, we included a user-name generator which is a link instead of button.

Are you referring to the "Learn more" circle thing? If so, link is probably appropriate, as it ends up opening the browser / a webpage

File: src/scss/list.scss [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ﻿@import "variables.scss";
 
-.list > a {
+.list > a,
+.list > button {
   display: block;
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply src/scss/list.scss:4]: Removed the duplicate unnecessary selectors (also in `vault.scss`). Will look into the `position: relative` issue

File: src/scss/base.scss [hunk 2]
```diff
@@ -50,7 +50,8 @@ img {
   border: none;
 }
 
-a {
+a,
+button {
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply src/scss/base.scss]: good spot. removing this fixes the "x" in the toasts, and seems to have no further adverse effect on my changes. thanks

File: src/scss/list.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply src/scss/list.scss]: Removed the `position: relative` and that indeed fixes the focus outline (on the cipher and send items)

**Comment by patrickhlauke** [comment]
Before/after for the small change made to the `:last-child` on action buttons.

Before:

Off-center outline around a 'Show' eye button
Off-center outline around a 'Copy' button

After:

Correctly centered outline around a 'Show' eye button
Correctly centered  outline around a 'Copy' button

**Comment by patrickhlauke** [comment]
Video (using NVDA screen reader on Windows) of the current issue with the "Options" expand/collapse control on the "Create send" view. Button lacks an accessible name (only announced as generic "Button") and doesn't announce its state (expanded or collapsed). It then shows the same button after this PR - it's now announced correctly as "Options", and announces whether it's expanded or collapsed.

https://user-images.githubusercontent.com/895831/166067851-621f8e61-a6a8-41b4-9db6-67b063c5a908.mp4

**Comment by patrickhlauke** [comment]
Videos showing the effect of adding all the various `aria-pressed` attributes to both the left-hand filters and the actual ciphers/sends in the second column. When they're active, they're now announced as "pressed". Otherwise, "not pressed". They are all announced as "toggle button"s.

https://user-images.githubusercontent.com/895831/166068292-59aeacb0-d0ed-415a-9f1d-f1b1f327eb73.mp4

https://user-images.githubusercontent.com/895831/166068308-b63a4ea5-5bf8-4cdd-bc51-e9f99be2802b.mp4

**Comment by patrickhlauke** [comment]
two oddities i noticed while testing this - but they happen even on `master`, so nothing related to any changes here:

there's an empty radio button in the password generator's username options

empty radio button

the list of sends always looks broken to start with, with empty/inoperable bits where previously defined sends would be. if i create new ones, they show up fine, until I restart the app and then they also show as broken.

broken sends

if I fire up the stable desktop app, they show up correctly though.

correct list of sends in stable app

**Comment by patrickhlauke** [comment]
@djsmith85 I think I've tackled all the things you mentioned (modulo that link in the generator, which - if it's the "Learn more" one - is fine)

**Review by djsmith85 — APPROVED** [review]
This is looking great and thank you for the visual examples. 

I've been unable to repro the issue with broken sends, but will keep an eye on it.

The empty radio button will be fixed once https://github.com/bitwarden/desktop/pull/1511 is merged.

The "Learn more" should stay as link, as you correctly said, it links to a webpage.

Thank you very much for tackling this 👍 

I've just realized that some change to the left hand navbar are happening here: https://github.com/bitwarden/desktop/pull/1487

@addisonbeck: I hope these don't conflict or if necessary you could work the changes from Patrick into your work if need be.

**Comment by patrickhlauke** [comment]
great stuff, thank you

**Pull request merged**